### PR TITLE
Callstack fix and refactors

### PR DESCRIPTION
**Pull request by bjourne**
This pr contains a bunch of refactorings to the vm code. Like I've removed some header files which I thought were to small, changed a lot of visitor structs to c++14-lambda's and made some methods into functions instead. Factor crashed on some callstack overflow errors instead of handling them correctly. It seems to have been caused by a to small stack reservation so I changed it from 1024 to 4096 so give it enough room.

### CODE REVIEW

**Comment by mrjbq7** [comment]
```
vm/debug.cpp:448:7: error: no member named 'unlock_console' in
      'factor::factor_vm'
  vm->unlock_console();
  ~~  ^
vm/debug.cpp:449:7: error: no member named 'handle_ctrl_c' in
      'factor::factor_vm'
  vm->handle_ctrl_c();
  ~~  ^
2 errors generated.
make[2]: *** [vm/debug.o] Error 1
make[1]: *** [macosx-x86-64] Error 2
make: *** [default] Error 2
```

File: vm/contexts.hpp [hunk 1]
```diff
@@ -12,7 +12,11 @@ enum context_object {
   OBJ_IN_CALLBACK_P,
 };
 
-static const cell stack_reserved = 1024;
+/* When the callstack fills up (e.g by to deep recursion), a callstack
+   overflow error is triggered. So before continuing executing on it
+   in general_error(), we chop of this many bytes to have some space
```
Comment (mrjbq7) [inline vm/contexts.hpp]: ``s/of/off/``

**Comment by mrjbq7** [comment]
I can fix those small things and merge or wait for you to push a patch?

**Comment by mrjbq7** [comment]
```diff
diff --git a/vm/contexts.hpp b/vm/contexts.hpp
index 20d773d..fc16981 100644
--- a/vm/contexts.hpp
+++ b/vm/contexts.hpp
@@ -14,7 +14,7 @@ enum context_object {
 
 /* When the callstack fills up (e.g by to deep recursion), a callstack
    overflow error is triggered. So before continuing executing on it
-   in general_error(), we chop of this many bytes to have some space
+   in general_error(), we chop off this many bytes to have some space
    to work with. */
 static const cell stack_reserved = 4096;
 
diff --git a/vm/debug.cpp b/vm/debug.cpp
index 46b4b44..43ec279 100644
--- a/vm/debug.cpp
+++ b/vm/debug.cpp
@@ -445,8 +445,8 @@ void factor_vm::factorbug_usage(bool advanced_p) {
 }
 
 static void exit_fep(factor_vm* vm) {
-  vm->unlock_console();
-  vm->handle_ctrl_c();
+  unlock_console();
+  handle_ctrl_c();
   vm->fep_p = false;
 }

```

**Comment by mrjbq7** [comment]
Tests look good, decided to merge and committed fix in 5c0d8c7e4fe2479bba5e33ba5abe013d544e2424.

**Comment by bjourne** [comment]
Great, your changes is right. :+1:  I don't understand how I missed that.

**Comment by mrjbq7** [comment]
No worries!!

**Pull request closed without merging**